src/pages/dashboard/activities/SettingsActivity.tsx

```diff
@@ -401,15 +402,61 @@ const PrimeManage: VoidComponent<{ dongleId: string }> = (props) => {
 
 const DeviceSettingsForm: VoidComponent<{ dongleId: string; device: Resource<Device> }> = (props) => {
   const [deviceName] = createResource(props.device, getDeviceName)
-
+  const [deviceUsers, { refetch: refetchDeviceUsers }] = createResource(props.dongleId, getDeviceUsers)
+  const [shareEmail, setShareEmail] = createSignal('')
   const [unpair, unpairData] = useAction(async () => {
     const { success } = await unpairDevice(props.dongleId)
     if (success) window.location.href = window.location.origin
   })
+  const [share, shareData] = useAction(async () => {
+    const { success } = await grantDeviceReadPermission(props.dongleId, shareEmail())
+    if (success) refetchDeviceUsers()
+    setShareEmail('')
+  })
+
+  const [unshareLoading, setUnshareLoading] = createSignal(false)
+
+  const unshare = async (email: string) => {
+    setUnshareLoading(true)
+    const { success } = await removeDeviceReadPermission(props.dongleId, email)
+    if (success) refetchDeviceUsers()
+    setUnshareLoading(false)
+  }
 
   return (
     <div class="flex flex-col gap-4">
       <h2 class="text-lg">{deviceName()}</h2>
+      <Show when={props.device()?.is_owner}>
+        <div class="flex flex-col gap-2">
+          <h3 class="text-md">{(deviceUsers() || []).length - 1 > 0 ? 'shared with:' : 'share device'}</h3>
+          <For each={deviceUsers()} fallback={<div>loading</div>}>
+            {(user, _index) => (
+              <Show when={user.permission !== 'owner'}>
+                <div class="flex items-center gap-2 justify-between">
+                  <div>{user.email}</div>
+                  <Button color="error" onClick={() => unshare(user.email)} loading={unshareLoading()}>
+                    <Icon name="delete" />
+                  </Button>
+                </div>
+              </Show>
+            )}
+          </For>
+          <div class="flex items-center gap-2 justify-between">
+            <TextField
```

Comment:
Make the text field fill the entire space instead of leaving a gap between it and the share button